UserspaceEmulator: Implement a dedicated FPU

Hendiadyoin1 opened the pull request:
this is to separate it form the CPU, because it uses it own state machine
also this is more faithful to the original hardware implementation

this adds more support for more FloatingPoint control and comparisons and nearly all of the MMX extension

the current current roadblocks are: 
* ambiguity of some instructions with maybe implied arguments, that we do not handle separately  and I therefore don't know how to handle them
* it's very difficult to find out things like "rounding has happened" without loading the actual state word and operating in assembly
* there is no good way to differentiate QNaN from SNaN as far as I know

Still missing instructions:
* FBLDB
* FBSSTP
* FXTRACT
* FCOM[P]
* FUCOM[P[P]]
* FNENI
* FNDISI
* FNSETPM
* FPUNPCK*B // gcc wants weird types, and I was fed up with the 3 lines of casting
* MOVQ (long-mode)
Unimplemented Features:
* Rounding from fp to fp
* Full ShadowValue support
* SNan/QNaN support
* FPU-precision control
* Exception handling

Known Issues:
* LibM test:
  * hyperbolics are broken (similar in upstream)
  * gamma is really broken -> crash (it can go way further than upstream)
* 3DFileViewer looks.... empty (same on master)
* Cube-trinagles clip weird (same on master)
* Not full test coverage (please don't force me to do that)

awesomekling reviewed (changes requested):
Please spell check your code, comments and commit messages.
Also, please wrap your commit messages at 72 characters.

Hendiadyoin1 commented:
Just fixed the few spelling mistakes I could find and fixed them, tell me if I missed any

also the maximum line length used in the commit messages is 66, and this is checked by the ci checks

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -0,0 +1,993 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2020, Leon Albrecht <loen2002.la@gmail.com>
```
It's 2021 now.

Hendiadyoin1 replied:
oh.. And apparently I can't spell my mail correctly either

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
CONTROL

Hendiadyoin1 replied:
fixed

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
TRANSFER

Hendiadyoin1 replied:
fixed

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2020, Leon Albrecht <loen2002.la@gmail.com>
```
It's 2021 now.

Hendiadyoin1 replied:
see above

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
```diff
@@ -0,0 +1,603 @@
+        TRUNK = 0x11
+    };
+
+    // FIXME: can we make this more a constexpr?
```
This function doesn't look like a good `constexpr` candidate. What am I missing?

Hendiadyoin1 replied:
its only called with constexpr Arguments, so its technically possible to optimize the if statements away, would ALWAYS_INLINE solve that?

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
```diff
@@ -0,0 +1,603 @@
+        m_fpu_tw = 0;
+    }
+
+    // FIXME: better MMx shadow checks
```
MMX

Hendiadyoin1 replied:
fixed

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
```diff
@@ -0,0 +1,603 @@
+        long double m_st[8] { 0 };
+        struct
+        {
+            // this should represents *packed* data not an int
```
represents => represent

Hendiadyoin1 replied:
fixed

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -26,6 +26,15 @@
 
 namespace UserspaceEmulator {
 
+template<typename T>
```
(In commit message. complience => compliance)

Hendiadyoin1 replied:
fixed

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -744,20 +744,76 @@ void SoftFPU::FSQRT(const X86::Instruction&)
 void SoftFPU::FXTRACT(const X86::Instruction&) { TODO_INSN(); }
 
 // COMPARISON
+
```
(In commit message: suport => support)

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -810,8 +866,66 @@ void SoftFPU::FUCOMIP(const X86::Instruction& insn)
+}
+void SoftFPU::FXAM(const X86::Instruction&)
+{
+    // noexcept!
```
What does this mean?

Hendiadyoin1 replied:
this is a notice, that that function cannot raise a FPU Exception, shall I reword it or delete it?

awesomekling commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -951,7 +957,13 @@ void SoftFPU::FPTAN(const X86::Instruction&)
-void SoftFPU::FPATAN(const X86::Instruction&) { TODO_INSN(); }
+void SoftFPU::FPATAN(const X86::Instruction&)
+{
+    // FIXME: set C1  on stackunderflow, or on rounding
```
stackunderflow => stack underflow

Hendiadyoin1 replied:
fixed

awesomekling reviewed (changes requested):
>also the maximum line length used in the commit messages is 66, and this is checked by the ci checks
or do you mean, I should wrap later in some cases?

Yes, please don't insert random line breaks in your commit messages and comments.
It makes your commits look different from everyone else's commits and it's much harder to read.

Also, please try to write both commit messages and comments with proper punctuation.
Your code changes are nice, but please make an effort with your human language as well. :)

Hendiadyoin1 commented:
Ok after the other commit got merged I have to rebase, with that I'll sit down and implement the  new MMX instructions

awesomekling commented:
> Human languange is difficult... (english is really inconsistent with its writing system)

Please use a spell checker.

> In the first commit I chose to put the line break early, because otherwise it would be in the middle of the second part of the sentence and therefore unpleassant to read

Please format your commit messages in accordance with project style.

Hendiadyoin1 commented:
@awesomekling I added MMX support now, I hope all commits are up to the standard and there aren't to many spelling mistakes;
If I need to squash things down, tell me, I know its a lot of small-ish committs.

Hendiadyoin1 commented:
Sadly these still aren't tested, due to the same bug as before

Hendiadyoin1 commented:
Now being able to test, I will need some time to verify all this...

awesomekling commented:
> @awesomekling I added MMX support now, I hope all commits are up to the standard and there aren't to many spelling mistakes;

Per the [contribution guidelines](https://github.com/SerenityOS/serenity/blob/773c17b6a0cb6cac1a40fccb6f285fe027497ff7/CONTRIBUTING.md):

>Write your commit messages in proper English, with care and punctuation.

The first commit message in this PR looks like this:

```
UserspaceEmulator: Start implementation of an FPU

this is to separate it form the cpu, because it uses its
own state machine

also this is more faithful to the original hardware implementation
and may make the code better structured for future iterations

also this adds the basics for a future mmx implementation
```

Here's what it would look like with spelling and punctuation fixed up:

```
UserspaceEmulator: Start implementation of an FPU

This is to separate it from the CPU, because it uses its own state
machine.

Also this is more faithful to the original hardware implementation
and may make the code better structured for future iterations.

Also this adds the basics for a future MMX implementation.
```

Please be respectful of other people's time. We put effort into our commit messages because thousands of people read them. :)

Dexesttp reviewed (commented):
I haven't dove into the code at all, but this looks fairly readable if tedious. It's a bit of a shame that the "Start implementation of a FPU" commit include both code move to a new file and new code added instead of being split in two, and that some commits go back-and-forth on the function declarations ("Inline all FPU helpers") or the mod() / m_rm bytes ("Add MMX Instructions " and "Cleanup use of ModRM byte"), but at this point it feels like it would be a hassle to rework the history and include them directly.

Here's some little things that I saw.

sin-ack commented on Userland/Libraries/LibC/stdio.cpp:
```diff
@@ -657,6 +657,11 @@ char* fgets(char* buffer, int size, FILE* stream)
     return ok ? buffer : nullptr;
 }
 
+char* gets(char* buffer)
+{
+    return fgets(buffer, -1, stdout);
+}
+
```
Why is this even needed?

Hendiadyoin1 replied:
I needed this for a programm that I used for testing, that had questionable Copyrights attached though and failed the ci, so I removed the test, but left the patches

shall I make this another pull request?

sin-ack commented on AK/Format.cpp:
```diff
@@ -374,6 +374,53 @@ void FormatBuilder::put_f64(
     put_string(string_builder.string_view(), align, min_width, NumericLimits<size_t>::max(), fill);
 }
+
+void FormatBuilder::put_f80(
+    long double value,
+    u8 base,
+    bool upper_case,
+    Align align,
+    size_t min_width,
+    size_t precision,
+    char fill,
+    SignMode sign_mode)
```
This is just `put_f64` with some types changed to `long double`. Is there a better way to do this that doesn't require copying code?

Hendiadyoin1 replied:
I can make it a
```cpp
template<FloatingPoint T>
void FormatBuilder::put_float(
    T value,
    ...
)
{
    ...
}
```
instead, that might be a better idea even

Hendiadyoin1 replied:
GCC does not seem very cooperative today,
If you get it working, just make a pull request to my repo and I will merge it then

sin-ack commented on AK/Format.cpp:
```diff
@@ -568,6 +615,29 @@ void Formatter<bool>::format(FormatBuilder& builder, bool value)
     }
 }
 #ifndef KERNEL
+void Formatter<long double>::format(FormatBuilder& builder, long double value)
```
Same as `put_f80`.

sin-ack commented on Userland/Libraries/LibX86/Types/u256.h:
```diff
@@ -147,6 +147,11 @@ class u256 {
     }
 
     // bitwise
+    ALWAYS_INLINE constexpr u128 operator~() const
```
This should probably return a `u256` :^)

Hendiadyoin1 replied:
That is indeed the case, will be fixed when I get to coding next

sin-ack commented on Userland/Libraries/LibX86/Types/Formatter.cpp:
```diff
@@ -49,12 +50,13 @@ void AK::Formatter<u128>::format(AK::FormatBuilder& builder, u128 value)
-    if (m_width.value() > lower_length) {
-        builder.put_u64(value.high(), base, m_alternative_form, upper_case, m_zero_pad, m_align, m_width.value() - lower_length, m_fill, m_sign_mode);
-        builder.put_u64(value.low(), base, false, upper_case, m_zero_pad, m_align, m_width.value(), m_fill, m_sign_mode);
+    u16 lower_length = sizeof(u64) * 8 / base;
+    auto width = m_width.value_or(0);
+    if (value.high()) {
+        builder.put_u64(value.high(), base, m_alternative_form, upper_case, m_zero_pad, m_align, max(width - lower_length, 0ul), m_fill, m_sign_mode);
+        builder.put_u64(value.low(), base, false, upper_case, m_zero_pad, m_align, width, m_fill, m_sign_mode);
```
Please explain what exactly this fixes in the commit message. "Fix" doesn't explain what was fixed and why.

Hendiadyoin1 replied:
So elaborate in the comit message?
Ok will do.

Will be something like:
> In the past we didn't calculate the part_length correctly which led to us only printing the lower half of the Value,
also we took the length argument as granted, which is more than often not the case,
in addition to that the lenght argument isn't treated as a upperbound anymore

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftCPU.h:
```diff
@@ -1153,7 +1163,10 @@ class SoftCPU final
     void update_code_cache();
 
 private:
+    friend class SoftFPU;
```
`class` isn't needed here because you include `"SoftFPU.h"`.

Hendiadyoin1 replied:
Good to know

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -0,0 +1,994 @@
+ALWAYS_INLINE void warn_if_uninitialized(T value_with_shadow, const char* message)
+{
+    if (value_with_shadow.is_uninitialized()) [[unlikely]] {
+        reportln("\033[31;1mWarning! Use of uninitialized value: {}\033[0m\n", message);
```
Usage of ANSI escapes directly in `reportln` feels a little bit iffy.

Hendiadyoin1 replied:
It the same in the SoftCPU, this is a carbon copy of that...

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -0,0 +1,994 @@
+
+long double SoftFPU::fpu_get(u8 index) const
+{
+    VERIFY(index < 8);
```
Can the 8 be put into a `constexpr` variable instead of being a magic number?

Hendiadyoin1 replied:
I have conflicting feelings about that, this limit is the size of the stack and should always be 8 and making it a variable, even a constexpr one will make it seem like it can be changed, which is not possible on a hardware level - there are only 3 bits reserved to index into this- so maybe not?

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
`fpu_is_set` sounds a bit awkward. How about `has_value_at`?

Hendiadyoin1 replied:
It can have a value even if it isn't set though...

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
```diff
@@ -0,0 +1,665 @@
+class Emulator;
+class SoftCPU;
+
+class SoftFPU final {
```
Almost everything in this class is `fpu_` prefixed. You are already defining the methods in the `SoftFPU` class, you do not need to prefix them.

Hendiadyoin1 replied:
this is to differentiate the x87 part and the mmx part of this class, so I see your point, but there is a reason for it though

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
The comments are screaming at me. Please use normal casing.

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
This comment seems unnecessary.

sin-ack commented on Userland/Libraries/LibX86/Instruction.cpp:
```diff
@@ -592,7 +592,7 @@ static void build_slash_reg(u8 op, u8 slash, const char* mnemonic, InstructionFo
 
     build_slash(0xDF, 0, "FILD", OP_FPU_RM32, &Interpreter::FILD_RM16);
-    build_slash_reg(0xDF, 0, "FFREEP", OP_FPU_reg, &Interpreter::FFREEP);
+    build_slash_reg(0xDF, 0, "FFREEP", OP_FPU_reg, &Interpreter::FFREEP); // undocumented
```
This comment seems unnecessary.

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.h:
Uh... is this clang-format's doing? :sweat_smile: This looks a bit weird, can we unroll this with a `using`?

Hendiadyoin1 replied:
I formatted it this way...
and the reason for it looking weird is because it is weird.... (thank gcc for that)
> long doubles are weird with their supposed sizes

Hendiadyoin1 replied:
>  can we unroll this with a using?

idk, how would you do it?

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -0,0 +1,994 @@
+
+void SoftFPU::fpu_push(long double value)
+{
+    if (fpu_is_set(7))
```
More magic numbers. Something like `SoftFPU::STACK_TOP - 1` would be nice.

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -107,6 +109,59 @@ long double SoftFPU::fpu_pop()
     return ret;
 }
 
+void SoftFPU::fpu_set_exception(FPU_Exception ex)
```
Please squash these changes. There is no reason for this change to be separate from the initial FPU patch.

sin-ack commented on Userland/DevTools/UserspaceEmulator/SoftFPU.cpp:
```diff
@@ -1319,12 +1319,13 @@ void SoftFPU::PSRLQ_mm1_mm2m64(const X86::Instruction&) { TODO_INSN(); }
-void SoftFPU::MOVQ_mm1_m64_mm2(const X86::Instruction&) { TODO_INSN(); }
+void SoftFPU::MOVQ_mm1m64_mm2(const X86::Instruction&) { TODO_INSN(); }
+void SoftFPU::MOVQ_mm1_rm64(const X86::Instruction&) { TODO_INSN(); }; // long mode
+void SoftFPU::MOVQ_rm64_mm2(const X86::Instruction&) { TODO_INSN(); }; // long mode
```
This patch should also be a part of the initial FPU patch, since it doesn't do anything beyond stubbing out the MMX instructions.

Hendiadyoin1 replied:
I think I should separate changes to different parts of the code as separated as possible, and this is a change Mainly for LibX86 and does not fully concern the Emulator

sin-ack reviewed (changes requested):
Great work for such a big change, but it needn't be split across so many commits. Many of the commits just build on the initial FPU commit and should be squashed into a single commit that splits the CPU from the FPU.

Hendiadyoin1 commented on Userland/Libraries/LibM/math.h:
```diff
@@ -29,6 +29,21 @@ __BEGIN_DECLS
 #define MAXFLOAT FLT_MAX
 
+// ruthlessly copied these from glibc
+#define M_El 2.718281828459045235360287471352662498L        /* e */
```
See comment...

I didn't really notice them, I can remove them whe I get to working on it again.

Hendiadyoin1 commented:
After the rebase one of the tests seems uncooperative, and the test seems to fail on native mashines too (Works for me with `-O0 -m32` and fails with `-O1 -m32` and higher, failed on @gunnarbeutner's machine on default settings (`-m32`))
the test in question is the environment store and load test

nico commented:
FWIW I looked this over and it looks roughly ok to me (I didn't look in detail).

I guess this would be easier to merge if it was several pull requests (eg one that just moves stuff to a new file without any other changes, and then most of the rest in a separate one).

Hendiadyoin1 commented:
This will be closed in favor of #8610
